lib/puma/state_file.rb

```diff
@@ -8,8 +8,9 @@ def initialize
       @options = {}
     end
 
-    def save(path)
+    def save(path, permission = nil)
       File.write path, YAML.dump(@options)
+      File.chmod(permission, path) if permission
```

Comment:
There's a race condition here: the file is created possibly insecure and then fixed. A more secure way is to ensure permissions before writing. For example (untested):
```suggestion
      if permission
        FileUtils.touch(path)
        File.chmod(permission, path)
      end
      File.write path, YAML.dump(@options)
```

Reply:
I believe you can do this in a single transaction block that might be a bit more clean:

```
File.open(path) do |file|
  file.chmod(permission) if permission
  file.write(YAML.dump(@options))
end
```